### PR TITLE
[Service Catalog] render provision form using broker-provided schema

### LINKED ISSUE #239: Use optional provision/bind JSON scehma from Service Catalog
The [Open Service Broker API](https://github.com/openservicebrokerapi/servicebroker/) allows brokers to provide a JSON schema that defines provisioning/update/bind parameters. It would be very nice if Kubeapps could leverage that to provide a nicer screen for creating services instead of simply providing a JSON blob. OpenShift does this and I believe that the Cloud Foundry Apps Manager does as well. This is _optional_ however, so there should be a good fallback. 

Here is an example of the UI in OpenShift:

### PR DESCRIPTION
**Pull request by prydonius**
fixes #239

cc @sameersbn @jeremyrickard @mhrivnak

screen shot 2018-06-05 at 18 21 10

screen shot 2018-06-05 at 18 20 54

### CODE REVIEW

**Comment by mhrivnak** [comment]
:+1:

File: dashboard/src/components/ProvisionButton/index.tsx [hunk 1]
```diff
@@ -173,62 +116,40 @@ class ProvisionButton extends React.Component<IProvisionButtonProps, IProvisionB
...
+    const { name, parameters, ...rest } = formData;
     if (selectedClass && selectedPlan) {
       const provisioned = await provision(
-        releaseName,
+        name,
         namespace,
         selectedClass.spec.externalName,
         selectedPlan.spec.externalName,
-        parametersObject,
+        parameters || rest,
```
Comment (andresmgot) [inline dashboard/src/components/ProvisionButton/index.tsx]: out of curiosity, might be cases in which that the hash `parameters` and other properties are set (...rest)?

File: dashboard/src/components/ProvisionButton/index.tsx [hunk 2]
```diff
@@ -46,34 +39,40 @@ const RequiredRBACRoles: IRBACRole[] = [
...
-    const { plans, classes } = this.props;
-    const { selectedClass, selectedPlan } = this.state;
+    const { selectedPlan } = this.props;
+    let schema = selectedPlan.spec.instanceCreateParameterSchema;
+    if (schema) {
+      schema.properties = {
+        name: NameProperty,
+        ...(schema.properties || {}),
+      };
+      schema.required = [...(schema.required || []), "name"];
```
Comment (andresmgot) [inline dashboard/src/components/ProvisionButton/index.tsx]: where are the required fields validated?

**Comment by jeremyrickard** [comment]
My overall familiarity with the kubeapps code base is a little limited, but this looks great as a new feature! Thanks for this @prydonius

File: dashboard/src/components/ProvisionButton/index.tsx (hunk 2 above)
Comment (prydonius) [reply dashboard/src/components/ProvisionButton/index.tsx]: For now this just uses the basic browser validations with the `required="true"` attribute on input elements.

File: dashboard/src/components/ProvisionButton/index.tsx (hunk 1 above)
Comment (prydonius) [reply dashboard/src/components/ProvisionButton/index.tsx]: `parameters` is used when the broker doesn't provide a schema (instanceCreateParameterSchema is null). It's possible that the broker schema defines a key called `parameters` which would conflict with this (probably what you're referring to). I don't think this is very likely, but we could rename this to `kubeappsParameters` or something a bit more unique to be more sure. WDYT?

Comment (andresmgot) [reply dashboard/src/components/ProvisionButton/index.tsx]: yes, it would be better to have `kubeappsParameters` just to be sure.

Comment (prydonius) [reply dashboard/src/components/ProvisionButton/index.tsx]: @andresmgot updated - PTAL!

**Pull request merged**